lib/IRGen/IRGenSIL.cpp

```diff
@@ -2472,6 +2476,76 @@ static void emitDynamicSelfMetadata(IRGenSILFunction &IGF) {
   IGF.setDynamicSelfMetadata(selfTy, isExact, value, selfKind);
 }
 
+namespace {
+// Notes the use of foreign types in generic arguments for C++ interop.
+class NoteUseOfForeignTypeMetadata
+    : public TypeVisitor<NoteUseOfForeignTypeMetadata, void, bool> {
+public:
+  NoteUseOfForeignTypeMetadata(IRGenerator &IRGen) : IRGen(IRGen) {}
+
+  void visitType(TypeBase *, bool) {}
+
+  void visitTupleType(TupleType *TT, bool isInGenericArgument) {
+    for (auto type : TT->getElementTypes())
+      TypeVisitor::visit(type, isInGenericArgument);
+  }
+
+  void visitSugarType(SugarType *sugarTy, bool isInGenericArgument) {
+    TypeVisitor::visit(sugarTy->getSinglyDesugaredType(), isInGenericArgument);
+  }
+
+  void visitClassType(ClassType *CT, bool isInGenericArgument) {
+    noteUseOfTypeMetadata(CT->getDecl(), isInGenericArgument);
```

Comment:
For example, for non-generic nominal types you also have to visit getParent(), which might be generic (`struct G<T> { struct Nested {} }`). If that can't come up in your case, you can add an `ASSERT(!CT->isSpecialized())`

Reply:
Would `Type::walk()` also visit the parents?

Reply:
Yes, it will visit everything.